enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,492 @@
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+- This will match all newly attached AWS EBS PVs, just before kubelet formats them.
+- This will match all local disks that already have PVs, but these PVs are not bound / used yet.`
+
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce two new features in LSO
+  - Auto discovery of local devices
+    - This will involve the intoduction of two new CRs `LocalVolumeDiscovery` and `LocalVolumeNodeDiscoveryResult`.
+    - The purpose of this will be to expose local devices available in a node via the `LocalVolumeNodeDiscoveryResult` CR.
+    - The device discovery will be done once via a discovery job. If an update happens to the `LocalVolumeNodeDiscoveryResult` CR, the jobs gets retriggered
+  - Auto provisioning of localDevices
+    - This will involve introduction of a new CR called `AutoDetectVolume`.
+    - The pupose of this will be to auto discover and provision PVs on devices which match the inclusion filter present in the CR.
+    - This will involve a continous process of discovery of devices via the diskmaker daemons. Any discovered devices which matches the inclusion filter will be considered for provisioning of PVs.
+
+Once we have the detected devices the administrator can create the localPVs by explicitly selecting the disks. This can be done by the `localVolume` CR. The other option would be to create localPVs via the AutoDetectVolume CR and passing the inclusion filters in it.
+
+
+#### Workflow of LocalPV creation via LocalVolume CR
+
+- Discovery: The user can choose to run discovery if they want to understand what devices are available across the nodes in the cluster.
+    - ##### NOTE: This step is optional for users directly creating CRs. They might already know what CRs they need for step 2.
+- PV Creation: After the user has decided how to created PVs, they can choose to create PVs with either the LocalVolume CR or the AutoDetectVolume CR.
+
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscovery` and `LocalVolumeNodeDiscoveryResult` CR:
+
+```go
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
+
+type LocalVolumeDiscovery struct {
```

Comment:
For discussion:

LSO API group is `local.storage.openshift.io`.
* IMO we cut the `Local` from the struct name without any issue (everyone who sees `VolumeDiscovery.local.storage.openshift.io` must see that it's about *local* volumes).
* Can we cut `Volume` too? Is `Discovery.local.storage.openshift.io` descriptive enough? IMO it looks better.
* And at last, can we add Job (`DiscoveryJob` / `VolumeDiscoveryJob`) to emphasize it's one-shot operation? `...NodeDiscoveryResult` is result of the job, it matches nicely together.


In similar fashion, `LocalVolumeNodeDiscoveryResult` -> `NodeDiscoveryResult`? `local.storage.openshift.io` IMO already implies that the discovered stuff is *local storage*.

@gnufied, WDYT?

Reply:
I think I like it. shortening some these names will help remove the clutter.

Reply:
In my experience, the type names are most often without being fully qualified. There is no issue if the fully qualified name is used, but there can be collisions of types if short names are used. If we use a common name such as `Discovery`, if the same name is used in a separate group name (`Discovery.foo.io`), anyone referencing the type would be forced to use the fully qualified name to avoid the collision on the name `Discovery`. This is you see CRD types named such as `EtcdCluster` and `CephCluster` to avoid collision on `Cluster`. With this in mind, I would vote for the type names `LocalVolumeDiscoveryJob` and `LocalVolumeDiscoveryResult`. Seems like we don't need `Node` in the result CRD name.

Reply:
updated with `LocalVolumeDiscoveryJob` and `LocalVolumeDiscoveryResult`